drivers/counter/counter_ll_stm32_rtc.c

```diff
@@ -164,20 +192,39 @@ static int rtc_stm32_stop(const struct device *dev)
 }
 
 
-static uint32_t rtc_stm32_read(const struct device *dev)
+tick_t rtc_stm32_read(const struct device *dev)
 {
 #if !defined(COUNTER_NO_DATE)
 	struct tm now = { 0 };
 	time_t ts;
-	uint32_t rtc_date, rtc_time, ticks;
+	uint32_t rtc_date, rtc_time;
+	tick_t ticks;
+#ifdef CONFIG_COUNTER_RTC_STM32_SUBSECONDS_BASED
+	uint32_t rtc_subseconds;
+#endif /* CONFIG_COUNTER_RTC_STM32_SUBSECONDS_BASED */
 #else
 	uint32_t rtc_time, ticks;
 #endif
 	ARG_UNUSED(dev);
 
-	/* Read time and date registers */
-	rtc_time = LL_RTC_TIME_Get(RTC);
+	/* Read time and date registers. Make sure value of the previous register
+	 * hasn't been changed while reading the next one.
+	 */
 #if !defined(COUNTER_NO_DATE)
+	do {
+		rtc_date = LL_RTC_DATE_Get(RTC);
+
+#ifdef CONFIG_COUNTER_RTC_STM32_SUBSECONDS_BASED
+		do {
+			rtc_time = LL_RTC_TIME_Get(RTC);
+			rtc_subseconds = LL_RTC_TIME_GetSubSecond(RTC);
+		} while (rtc_time != LL_RTC_TIME_Get(RTC));
+#else /* CONFIG_COUNTER_RTC_STM32_SUBSECONDS_BASED */
+		rtc_time = LL_RTC_TIME_Get(RTC);
+#endif
+
+	} while (rtc_date != LL_RTC_DATE_Get(RTC));
```

Comment:
Reference manual says : "To ensure consistency between the 3 values, reading either RTC_SSR or RTC_TR locks the values in the higher-order calendar shadow registers until RTC_DR is read".
So imho you must call LL_RTC_TIME_GetSubSecond() before LL_RTC_TIME_Get() and LL_RTC_DATE_Get().

Reply:
Thanks for the info, I will take a look one more time.

Reply:
Well actually as LL_RTC_EnableShadowRegBypass is called this won't be true. Even if bypass was not enabled it won't work for the STM32F2 as @gautierg-st pointed out in #62304.

Reply:
True, thanks for the point.